### PR TITLE
WIP:Query Offset

### PR DESCRIPTION
**Pull request by Yattol**
https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/blob/7e9b3d8d7eed9cf45e6afa941b3f20feb8f02011/README.md#track-offset
Waiting for consumer to be developed. Tests still to be done

### CODE REVIEW

File: test/e2e/store_offset.test.ts [hunk 1]
```diff
@@ -30,10 +30,14 @@ describe("declare consumer", () => {
       (message: Message) => messages.push(message.content)
     )
 
-    await consumer.storeOffset(Offset.offset(1n).value!)
+    await consumer.storeOffset(1n)
```
Comment (gpad) [inline test/e2e/store_offset.test.ts:33]: This test should be changed:
1. remove the skip
2. store the offset while consuming
3. Verify with the query we have stored everything

File: test/e2e/query_offset.test.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  it("the consumer is able to track the offset of the stream through queryOffset method", async () => {
+    const messages: Buffer[] = []
+    const publisher = await connection.declarePublisher({ stream: testStreamName })
+    await publisher.send(Buffer.from("hello"))
+    const consumer = await connection.declareConsumer(
+      { stream: testStreamName, offset: Offset.first(), consumerRef: "my_consumer" },
+      (message: Message) => messages.push(message.content)
+    )
+    await consumer.storeOffset(1n)
```
Comment (gpad) [inline test/e2e/query_offset.test.ts:34]: Move the consumer inside the consumer

File: test/e2e/query_offset.test.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
+import { expect } from "chai"
+import { Connection } from "../../src"
+import { Message } from "../../src/producer"
+import { Offset } from "../../src/requests/subscribe_request"
+import { Rabbit } from "../support/rabbit"
+import { expectToThrowAsync, password, username } from "../support/util"
+import { createConnection } from "../support/fake_data"
+
+describe("declare consumer", () => {
```
Comment (gpad) [inline test/e2e/query_offset.test.ts:9]: The name is incorrect

File: test/e2e/query_offset.test.ts
Comment (gpad) [inline test/e2e/query_offset.test.ts:1]: Why don't merge this file the test/e2e/store_offset.test.ts ?

File: src/connection.ts
Comment (gpad) [inline src/connection.ts:1]: When we declare a consumer, I suggest adding a field to automatically retrieve the offset and start to use it when we publish a message without an explicit number. this feature is present in the C# client.

**Review by gpad — COMMENTED** [review]
it.skip("saving the store offset of a stream correctly",

File: test/e2e/query_offset.test.ts (hunk 2 above)
Comment (l4mby) [reply test/e2e/query_offset.test.ts:34]: Sorry, you want me to move the storeOffset where?

File: src/connection.ts
Comment (l4mby) [reply src/connection.ts:1]: Sorry so when this filed is set, we need to query the offset and then store it in the consumer?

Comment (gpad) [reply src/connection.ts:1]: Mmh, I think I have confused the offset of the consumer with the publiserId of the Producer forget it...

File: src/connection.ts [hunk 4]
```diff
@@ -299,8 +312,8 @@ export class Connection {
     return this.send(new CreditRequest({ ...params }))
   }
 
-  public storeOffset(params: StoreOffsetParams): Promise<void> {
-    return this.send(new StoreOffsetRequest(params))
+  public async storeOffset(params: StoreOffsetParams): Promise<void> {
+    return await this.send(new StoreOffsetRequest(params))
```
Comment (gpad) [inline src/connection.ts:316]:
```suggestion
  public  storeOffset(params: StoreOffsetParams): Promise<void> {
    return  this.send(new StoreOffsetRequest(params))
```

File: src/consumer.ts [hunk 5]
```diff
@@ -32,4 +32,10 @@ export class Consumer {
     if (!this.consumerRef) throw new Error("ConsumerReference must be defined in order to use this!")
     return this.connection.storeOffset({ stream: this.stream, reference: this.consumerRef, offsetValue })
   }
+
+  public async queryOffset(): Promise<bigint> {
+    if (!this.consumerRef) throw new Error("ConsumerReference must be defined in order to use this!")
+    const offset = await this.connection.queryOffset({ stream: this.stream, reference: this.consumerRef })
+    return offset
+  }
```
Comment (gpad) [inline src/consumer.ts:40]:
```suggestion
  public queryOffset(): Promise<bigint> {
    if (!this.consumerRef) throw new Error("ConsumerReference must be defined in order to use this!")
    return this.connection.queryOffset({ stream: this.stream, reference: this.consumerRef })
  }
```

File: test/e2e/offset.test.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+  describe("store", () => {
+    it("saving the store offset of a stream correctly", async () => {
+      let offset: bigint = 0n
+      const consumer = await connection.declareConsumer(
+        { stream: testStreamName, consumerRef: "my consumer", offset: Offset.next() },
+        async (message: Message) => {
+          await consumer.storeOffset(message.offset!)
+          offset = message.offset!
+          console.log(offset)
```
Comment (gpad) [inline test/e2e/offset.test.ts:34]: You can remove the `console.log`

File: test/e2e/offset.test.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        }
+      )
+      const publisher = await connection.declarePublisher({ stream: testStreamName })
+      await publisher.send(Buffer.from("hello"))
+      await publisher.send(Buffer.from("world"))
+      await eventually(async () => {
+        const result = await consumer.queryOffset()
+        expect(result).eql(offset)
+      })
+    }).timeout(10000)
```
Comment (gpad) [inline test/e2e/offset.test.ts:44]: Keep the AAA in this test

File: test/e2e/offset.test.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
```
Comment (gpad) [inline test/e2e/offset.test.ts:79]: AAA

File: test/e2e/offset.test.ts
Comment (gpad) [inline test/e2e/offset.test.ts:1]: Remove all the console.log

**Review by gpad — APPROVED** [review]

**Pull request merged**